pkg/repository/project-member.go

```diff
@@ -0,0 +1,196 @@
+	role := ProjectRoleFromMap(req.Member.Meta.Annotations)
+
+	// TODO: currently the API defines that only owners can update members so there is no possibility to elevate permissions
+	// probably, we should still check that no elevation of permissions is possible in case we later change the API
+
+	lastOwner, err := t.checkIfMemberIsLastOwner(ctx, membership)
+	if err != nil {
+		return connect.NewError(connect.CodeInternal, err)
+	}
+
+	if lastOwner && role != apiv2.ProjectRole_PROJECT_ROLE_OWNER {
+		return connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("cannot demote last owner's permissions"))
+	}
+
+	return nil
+}
+
+func (t *projectMemberRepository) validateDelete(ctx context.Context, req *mdcv1.ProjectMember) error {
+	lastOwner, err := t.checkIfMemberIsLastOwner(ctx, req)
+	if err != nil {
+		return connect.NewError(connect.CodeInternal, err)
+	}
+	if lastOwner {
+		return connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("cannot remove last owner of a project"))
+	}
+
+	return nil
+}
+
+func (t *projectMemberRepository) convertToInternal(msg *apiv2.ProjectMember) (*mdcv1.ProjectMember, error) {
+	return &mdcv1.ProjectMember{
+		Meta: &mdcv1.Meta{
+			Id: msg.Id,
+			Annotations: map[string]string{
+				ProjectRoleAnnotation: msg.Role.String(),
+			},
+		},
+	}, nil
+}
+
+func (t *projectMemberRepository) convertToProto(e *mdcv1.ProjectMember) (*apiv2.ProjectMember, error) {
+	return &apiv2.ProjectMember{
+		Id:        e.TenantId,
+		Role:      ProjectRoleFromMap(e.Meta.Annotations),
+		CreatedAt: e.Meta.CreatedTime,
+	}, nil
+}
+
+func (t *projectMemberRepository) create(ctx context.Context, c *ProjectMemberCreateRequest) (*mdcv1.ProjectMember, error) {
+	resp, err := t.s.mdc.ProjectMember().Create(ctx, &mdcv1.ProjectMemberCreateRequest{
+		ProjectMember: &mdcv1.ProjectMember{
+			Meta: &mdcv1.Meta{
+				Annotations: map[string]string{
+					TenantRoleAnnotation: c.Role.String(),
+				},
+			},
+			TenantId:  c.MemberID,
+			ProjectId: t.scope.projectID,
+		},
+	})
+	if err != nil {
+		return nil, errorutil.Convert(err)
+	}
+
+	return resp.ProjectMember, nil
+}
+
+func (t *projectMemberRepository) delete(ctx context.Context, e *mdcv1.ProjectMember) error {
+	_, err := t.s.mdc.ProjectMember().Delete(ctx, &mdcv1.ProjectMemberDeleteRequest{
+		Id: e.Meta.Id,
+	})
+	if err != nil {
+		return errorutil.Convert(err)
+	}
+
+	return nil
+}
+
+func (t *projectMemberRepository) find(ctx context.Context, query *ProjectMemberQuery) (*mdcv1.ProjectMember, error) {
+	if query.MemberId == nil {
+		return nil, fmt.Errorf("member id must be specified")
+	}
+
+	memberships, err := t.list(ctx, query)
+	if err != nil {
+		return nil, errorutil.Convert(err)
+	}
+
+	switch len(memberships) {
+	case 0:
+		return nil, connect.NewError(connect.CodeNotFound, fmt.Errorf("tenant %s is not a member of project %s", *query.MemberId, t.scope.projectID))
+	case 1:
+		// fallthrough
```

Comment:
What is this comment for?

Reply:
Changed comment to `// noop`, which is probably clearer.